<a href="https://colab.research.google.com/github/jeremiahsnipes/motorcycle_risk_project/blob/main/SouthernCA_Analysis%26Models.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
# Motorcycle Prediction Project

import pandas as pd

crashes_url = "https://raw.githubusercontent.com/jeremiahsnipes/motorcycle_risk_project/main/data/socal_motorcycle_crashes_updated.csv"
registration_url = "https://raw.githubusercontent.com/jeremiahsnipes/motorcycle_risk_project/main/data/socal_registration.csv"

df_crashes = pd.read_csv(crashes_url)
df_registration = pd.read_csv(registration_url)

df_crashes.head(), df_registration.head()



(   ACCIDENT_YEAR  JURIS COLLISION_DATE  COLLISION_TIME  DAY_OF_WEEK  \
 0           2021   1942     2021-01-11            1440            1   
 1           2021   1942     2021-01-10            1153            7   
 2           2021   1942     2021-01-26            1400            2   
 3           2021   1942     2021-01-19            1330            2   
 4           2021   1933     2021-01-25             856            1   
 
    CHP_SHIFT INTERSECTION WEATHER_1 WEATHER_2 STATE_HWY_IND  ...  \
 0          5            N         A         -             N  ...   
 1          5            N         A         -             N  ...   
 2          5            Y         A         -             N  ...   
 3          5            Y         A         -             N  ...   
 4          5            N         A         -             N  ...   
 
   COUNT_MC_INJURED   LATITUDE   LONGITUDE       COUNTY         CITY  \
 0                0        NaN         NaN  LOS ANGELES  LOS ANGELES   
 1    

In [9]:
df_registration.columns

Index(['County', 'Year', ' Total Registrations ', 'Motorcycle Registration',
       'Auto Registration', 'Truck Registration', 'Other Registration',
       'Unnamed: 7', 'Unnamed: 8', 'Unnamed: 9', 'Unnamed: 10', 'Unnamed: 11',
       'Unnamed: 12', 'Unnamed: 13', 'Unnamed: 14'],
      dtype='object')

In [8]:
df_crashes.columns


Index(['ACCIDENT_YEAR', 'JURIS', 'COLLISION_DATE', 'COLLISION_TIME',
       'DAY_OF_WEEK', 'CHP_SHIFT', 'INTERSECTION', 'WEATHER_1', 'WEATHER_2',
       'STATE_HWY_IND', 'LOCATION_TYPE', 'RAMP_INTERSECTION', 'TOW_AWAY',
       'COLLISION_SEVERITY', 'NUMBER_KILLED', 'NUMBER_INJURED', 'PARTY_COUNT',
       'PRIMARY_COLL_FACTOR', 'PCF_CODE_OF_VIOL', 'PCF_VIOL_CATEGORY',
       'PCF_VIOLATION', 'PCF_VIOL_SUBSECTION', 'HIT_AND_RUN',
       'TYPE_OF_COLLISION', 'ROAD_SURFACE', 'ROAD_COND_1', 'ROAD_COND_2',
       'LIGHTING', 'CONTROL_DEVICE', 'CHP_ROAD_TYPE', 'MOTORCYCLE_ACCIDENT',
       'ALCOHOL_INVOLVED', 'STWD_VEHTYPE_AT_FAULT', 'CHP_VEHTYPE_AT_FAULT',
       'COUNT_SEVERE_INJ', 'COUNT_VISIBLE_INJ', 'COUNT_COMPLAINT_PAIN',
       'COUNT_MC_KILLED', 'COUNT_MC_INJURED', 'LATITUDE', 'LONGITUDE',
       'COUNTY', 'CITY', 'POINT_X', 'POINT_Y', 'IS_WEEKEND', 'TIME_OF_DAY',
       'SEVERE_FATAL'],
      dtype='object')

In [14]:
# Standardize crash dataset column names
df_crashes = df_crashes.rename(columns={
    "COUNTY": "County",
    "ACCIDENT_YEAR": "Year"
})

# Clean registration dataset (strip whitespace)
df_registration.columns = df_registration.columns.str.strip()

# Drop Unamed Columns
df_registration = df_registration.loc[:, ~df_registration.columns.str.contains("Unnamed")]

df_registration = df_registration.rename(columns={
    "County": "County",
    "Year": "Year"
})

In [16]:
df_merged = df_crashes.merge(
    df_registration,
    on=["County", "Year"],
    how="left"
)

df_merged.head()

,Year,JURIS,COLLISION_DATE,COLLISION_TIME,DAY_OF_WEEK,CHP_SHIFT,INTERSECTION,WEATHER_1,WEATHER_2,STATE_HWY_IND,...,POINT_X,POINT_Y,IS_WEEKEND,TIME_OF_DAY,SEVERE_FATAL,Total Registrations,Motorcycle Registration,Auto Registration,Truck Registration,Other Registration
0,2021,1942,2021-01-11,1440,1,5,N,A,-,N,...,-118.468544,34.071571,0,Midday,1,7989032.0,155065.0,6386830.0,1150957.0,296180.0
1,2021,1942,2021-01-10,1153,7,5,N,A,-,N,...,-118.388931,34.036381,1,Midday,0,7989032.0,155065.0,6386830.0,1150957.0,296180.0
2,2021,1942,2021-01-26,1400,2,5,Y,A,-,N,...,-118.344032,34.068905,0,Midday,1,7989032.0,155065.0,6386830.0,1150957.0,296180.0
3,2021,1942,2021-01-19,1330,2,5,Y,A,-,N,...,-118.299980,34.049068,0,Midday,0,7989032.0,155065.0,6386830.0,1150957.0,296180.0
4,2021,1933,2021-01-25,856,1,5,N,A,-,N,...,-118.343918,33.936653,0,Morning,0,7989032.0,155065.0,6386830.0,1150957.0,296180.0


In [17]:
df_merged.columns.tolist()


['Year',
 'JURIS',
 'COLLISION_DATE',
 'COLLISION_TIME',
 'DAY_OF_WEEK',
 'CHP_SHIFT',
 'INTERSECTION',
 'WEATHER_1',
 'WEATHER_2',
 'STATE_HWY_IND',
 'LOCATION_TYPE',
 'RAMP_INTERSECTION',
 'TOW_AWAY',
 'COLLISION_SEVERITY',
 'NUMBER_KILLED',
 'NUMBER_INJURED',
 'PARTY_COUNT',
 'PRIMARY_COLL_FACTOR',
 'PCF_CODE_OF_VIOL',
 'PCF_VIOL_CATEGORY',
 'PCF_VIOLATION',
 'PCF_VIOL_SUBSECTION',
 'HIT_AND_RUN',
 'TYPE_OF_COLLISION',
 'ROAD_SURFACE',
 'ROAD_COND_1',
 'ROAD_COND_2',
 'LIGHTING',
 'CONTROL_DEVICE',
 'CHP_ROAD_TYPE',
 'MOTORCYCLE_ACCIDENT',
 'ALCOHOL_INVOLVED',
 'STWD_VEHTYPE_AT_FAULT',
 'CHP_VEHTYPE_AT_FAULT',
 'COUNT_SEVERE_INJ',
 'COUNT_VISIBLE_INJ',
 'COUNT_COMPLAINT_PAIN',
 'COUNT_MC_KILLED',
 'COUNT_MC_INJURED',
 'LATITUDE',
 'LONGITUDE',
 'County',
 'CITY',
 'POINT_X',
 'POINT_Y',
 'IS_WEEKEND',
 'TIME_OF_DAY',
 'SEVERE_FATAL',
 'Total Registrations',
 'Motorcycle Registration',
 'Auto Registration',
 'Truck Registration',
 'Other Registration']

In [18]:
# Confirm Classes in Outcome Variable (COLLISION_SEVERITY = KABCO)
# - 1 = Fatal , 2 = Severe Injury , 3 = Visible Injury , 4 = Complaint of Pain
# 5 = Property Damage Only

df_merged['COLLISION_SEVERITY'].value_counts().sort_index()


,count
COLLISION_SEVERITY,
1,964
2,5033
3,8865
4,4859


In [19]:
# Percentages in Outcome Classes
df_merged['COLLISION_SEVERITY'].value_counts(normalize=True).sort_index()

,proportion
COLLISION_SEVERITY,
1,0.048882
2,0.255210
3,0.449521
4,0.246387


We see strong Class Imbalance since 4.89% is fatal, 25.52% is Severe Injury, 44.95% is Visible Injury, and 24.64% is Complaint of Pain - Note Class 5 Property Damage Only is not included.

In [23]:
# Predictor and Outcome List

predictors = [
    'Year', 'DAY_OF_WEEK', 'COLLISION_TIME', 'TIME_OF_DAY', 'IS_WEEKEND',
    'WEATHER_1', 'LIGHTING', 'ROAD_SURFACE', 'ROAD_COND_1',
    'STATE_HWY_IND', 'LOCATION_TYPE', 'INTERSECTION', 'RAMP_INTERSECTION',
    'CHP_ROAD_TYPE', 'CONTROL_DEVICE',
    'TYPE_OF_COLLISION', 'PRIMARY_COLL_FACTOR', 'PCF_CODE_OF_VIOL',
    'PCF_VIOL_CATEGORY', 'PCF_VIOLATION', 'PCF_VIOL_SUBSECTION',
    'HIT_AND_RUN', 'TOW_AWAY', 'PARTY_COUNT', 'ALCOHOL_INVOLVED',
    'STWD_VEHTYPE_AT_FAULT', 'CHP_VEHTYPE_AT_FAULT',
    'LATITUDE', 'LONGITUDE', 'County', 'CITY',
    'Total Registrations', 'Motorcycle Registration',
    'Auto Registration', 'Truck Registration', 'Other Registration'
]

outcome = 'COLLISION_SEVERITY'

The final predictor set reflects a systematic effort to construct a modeling framework that is both statistically valid and scientifically defensible. Several variables were intentionally removed because they represented direct measures of crash severity. Fields such as NUMBER_KILLED, NUMBER_INJURED, COUNT_SEVERE_INJ, COUNT_VISIBLE_INJ, COUNT_COMPLAINT_PAIN, COUNT_MC_KILLED, and COUNT_MC_INJURED quantify the injury outcomes of the crash itself. Including these variables would have introduced outcome leakage, allowing the model to infer severity directly from injury counts rather than from the environmental, behavioral, and roadway conditions that precede the crash. Removing these fields ensures that the model predicts severity based on causal or contributory factors rather than on variables that are deterministically linked to the outcome.
Other variables were excluded because they were redundant or offered limited analytical value. For example, WEATHER_2 was removed due to its sparse and often duplicative nature relative to WEATHER_1. In SWITRS data, secondary weather conditions are frequently missing or simply restate the primary condition, and retaining them would have increased dimensionality without improving predictive performance. Similarly, POINT_X and POINT_Y were omitted because they duplicate the information contained in LATITUDE and LONGITUDE. Eliminating redundant spatial fields reduces multicollinearity and simplifies downstream modeling. Variables such as JURIS and CHP_SHIFT were also removed because they tend to act as proxies for more interpretable predictors like county, city, or time of day. Their exclusion supports a cleaner, more interpretable model structure.
The resulting predictor set emphasizes variables that capture meaningful temporal, environmental, roadway, behavioral, and geospatial characteristics of each crash. These predictors represent conditions present before or during the collision and therefore align with the conceptual goal of modeling severity as a function of contributing factors rather than consequences. The inclusion of registration-based exposure variables further strengthens the analytical framework by contextualizing crash counts within the broader population of registered vehicles. Collectively, these decisions produce a predictor set that is methodologically sound, free of outcome leakage, and well-suited for rigorous statistical modeling and machine learning applications in crash‑severity research.


In [24]:
# Creating Clean Model Dataframe with Desired Variables

model_df = df_merged[predictors + [outcome]].copy()
model_df.head()

,Year,DAY_OF_WEEK,COLLISION_TIME,TIME_OF_DAY,IS_WEEKEND,WEATHER_1,LIGHTING,ROAD_SURFACE,ROAD_COND_1,STATE_HWY_IND,...,LATITUDE,LONGITUDE,County,CITY,Total Registrations,Motorcycle Registration,Auto Registration,Truck Registration,Other Registration,COLLISION_SEVERITY
0,2021,1,1440,Midday,0,A,A,A,H,N,...,NaN,NaN,LOS ANGELES,LOS ANGELES,7989032.0,155065.0,6386830.0,1150957.0,296180.0,1
1,2021,7,1153,Midday,1,A,A,A,H,N,...,NaN,NaN,LOS ANGELES,LOS ANGELES,7989032.0,155065.0,6386830.0,1150957.0,296180.0,3
2,2021,2,1400,Midday,0,A,A,A,H,N,...,NaN,NaN,LOS ANGELES,LOS ANGELES,7989032.0,155065.0,6386830.0,1150957.0,296180.0,2
3,2021,2,1330,Midday,0,A,A,A,H,N,...,NaN,NaN,LOS ANGELES,LOS ANGELES,7989032.0,155065.0,6386830.0,1150957.0,296180.0,4
4,2021,1,856,Morning,0,A,A,A,H,N,...,33.936798,-118.343903,LOS ANGELES,INGLEWOOD,7989032.0,155065.0,6386830.0,1150957.0,296180.0,4


In [25]:
# Assess for Missing Values

model_df.isna().sum().sort_values(ascending=False)

,0
RAMP_INTERSECTION,19170
ALCOHOL_INVOLVED,18247
LOCATION_TYPE,14135
PCF_VIOL_SUBSECTION,11627
LONGITUDE,6212
LATITUDE,6212
PCF_VIOLATION,963
CHP_VEHTYPE_AT_FAULT,359
TOW_AWAY,142
STATE_HWY_IND,12


In [27]:
# Handle Categorical Variable Missingness By inputting "Unkown"

cat_cols = [
    'RAMP_INTERSECTION', 'ALCOHOL_INVOLVED', 'LOCATION_TYPE',
    'PCF_VIOL_SUBSECTION', 'PCF_VIOLATION', 'CHP_VEHTYPE_AT_FAULT',
    'TOW_AWAY', 'STATE_HWY_IND'
]

for col in cat_cols:
    model_df[col] = model_df[col].fillna("Unknown")

In [28]:
# Handle Numberic Variables by Filling With Median for Lat/Long

num_cols = ['LATITUDE', 'LONGITUDE']

for col in num_cols:
    model_df[col] = model_df[col].fillna(model_df[col].median())


In this stage of the analysis, the objective is to characterize and appropriately treat missing data in a way that preserves the scientific integrity of the crash‑severity modeling process. Missingness is not simply a technical inconvenience; it reflects underlying data‑collection mechanisms, reporting practices, and structural features of the SWITRS system. For this reason, the first step is to examine the pattern and magnitude of missing values across all predictors. The variables exhibiting the highest levels of missingness—RAMP_INTERSECTION, ALCOHOL_INVOLVED, LOCATION_TYPE, and PCF_VIOL_SUBSECTION—are not missing at random. Their absence is tied to the nature of the crash event and the reporting protocol. For example, a large proportion of crashes do not occur on ramps, so the ramp indicator is legitimately absent. Alcohol involvement is frequently coded as “unknown” unless a test is administered, meaning that missingness reflects uncertainty rather than omission. Similarly, location type is only recorded when the officer explicitly classifies the site, and PCF subsections are only present when the violation has a definable subsection. Because these missing values carry substantive meaning, removing the affected rows would distort the underlying distribution of crash types and introduce bias. Instead, the scientifically appropriate approach is to encode these missing values as their own category, preserving the informational content of “unknown” or “not applicable” conditions.
Moderate missingness appears in the geospatial variables LATITUDE and LONGITUDE, which represent failed geocoding attempts rather than conceptual absence. These missing values do not reflect meaningful crash characteristics but rather limitations in the geospatial matching process. Given the size of the dataset and the fact that these failures are not systematically related to severity, imputing the median latitude and longitude is a defensible strategy. This approach maintains the spatial structure of the dataset without discarding a substantial number of observations. Alternatively, one could remove rows missing both coordinates, but the proportion is small enough that median imputation avoids unnecessary data loss while preserving analytical power.
Variables with low levels of missingness—such as PCF_VIOLATION, CHP_VEHTYPE_AT_FAULT, TOW_AWAY, and STATE_HWY_IND—are treated using mode imputation. These fields are categorical and their missingness is minimal, making the mode an appropriate and statistically stable replacement. This ensures that the dataset remains complete without materially altering the distribution of these predictors.
The final step is to apply a structured, reproducible missing‑data strategy aligned with best practices in crash‑severity research. Categorical variables with meaningful or systematic missingness are assigned an explicit “Unknown” category, preserving the interpretability and avoiding the introduction of bias through deletion. Numeric variables, particularly geospatial coordinates, are imputed using the median to maintain central tendency without being influenced by outliers. This dual‑strategy approach ensures that the dataset remains analytically robust, avoids outcome leakage, and retains the full informational value of the SWITRS reporting structure. With missingness fully addressed, the dataset is now suitable for encoding, scaling, exploratory analysis, and subsequent model development.
